### PR TITLE
Set accepted status for HTTPRoutes with Service parents

### PR DESCRIPTION
**Pull request by adleong**
We add support for setting the status field on HTTPRoute resources which have a Service as a parent_ref.  We also make a number of simplifications to the logic of how statuses are determined.  For each parent ref which is a Server or Service, we add a status.  This status will either be "accepted" or "no matching parent" depending on if that parent resource exists in the cluster.  No status condition is generated for parents of other kinds.

### CODE REVIEW

File: policy-controller/k8s/status/src/index.rs [hunk 1]
```diff
@@ -325,6 +320,39 @@ impl kubert::index::IndexNamespacedResource<k8s::policy::Server> for Index {
     // to handle resets specially.
 }
 
+impl kubert::index::IndexNamespacedResource<k8s::Service> for Index {
+    fn apply(&mut self, resource: k8s::Service) {
+        let namespace = resource.namespace().expect("Service must have a namespace");
+        let name = resource.name_unchecked();
+        let id = ResourceId::new(namespace, name);
+
+        self.services.insert(id);
```
Comment (mateiidavid) [inline policy-controller/k8s/status/src/index.rs:329]: Should we perhaps hold more information in the cache aside from the namespace and name? For example, routes [should not attach to headless services](https://gateway-api.sigs.k8s.io/geps/gep-1426/#allowed-service-types). Similarly, services that have an `ExternalName` type are not meant to be attached to either due to security concerns.

Do we want to consider this when setting the status?

**Review by mateiidavid — COMMENTED** [review]
Nice, this simplifies it a lot. Looks great.

File: policy-controller/k8s/status/src/index.rs (hunk 1 above)
Comment (adleong) [reply policy-controller/k8s/status/src/index.rs:329]: Ah, I wasn't aware of these restrictions.  However, they make perfect sense.  Let's tackle in a follow up.

File: policy-controller/k8s/status/src/index.rs [hunk 2]
```diff
@@ -337,3 +365,33 @@ pub(crate) fn make_patch(
...
+        type_: "Accepted".to_string(),
+    }
+}
+
+fn accepted() -> k8s::Condition {
+    k8s::Condition {
+        last_transition_time: k8s::Time(now()),
+        message: "".to_string(),
+        observed_generation: None,
+        reason: "Accepted".to_string(),
```
Comment (hawkw) [inline policy-controller/k8s/status/src/index.rs:393]: nit, take it or leave it: i would probably want to use a string constant for anything that shows up multiple times, to guard against future typos. not a big deal though.

File: policy-controller/src/index_list.rs
Comment (hawkw) [inline policy-controller/src/index_list.rs:1]: looks like this branch accidentally undoes a change from the previous PR --- does it need to be rebased?

**Review by hawkw — APPROVED** [review]
looks good to me, modulo that i think the branch needs a rebase

**Review by mateiidavid — APPROVED** [review]

File: policy-controller/k8s/status/src/index.rs [hunk 3]
```diff
@@ -43,8 +43,11 @@ pub struct Index {
     claims: Receiver<Arc<Claim>>,
     updates: UnboundedSender<Update>,
 
-    http_routes: HashMap<ResourceId, Vec<ParentReference>>,
+    /// Maps HttpRoute ids to a list of their parent refs, regardless of if
+    /// those parents have accepted the route.
```
Comment (hawkw) [inline policy-controller/k8s/status/src/index.rs:47]: tiny grammar nit, TIOLI
```suggestion
    /// Maps HttpRoute ids to a list of their parent refs, regardless of
    /// whether those parents have accepted the route.
```

File: policy-controller/k8s/status/src/index.rs [hunk 4]
```diff
@@ -325,6 +320,39 @@ impl kubert::index::IndexNamespacedResource<k8s::policy::Server> for Index {
...
+        let id = ResourceId::new(namespace, name);
+
+        self.services.remove(&id);
+
+        // If we're not the leader, skip reconciling the cluster.
+        if !self.claims.borrow().is_current_for(&self.name) {
+            tracing::debug!(%self.name, "Lease non-holder skipping controller update");
+            return;
+        }
+        self.reconcile();
```
Comment (hawkw) [inline policy-controller/k8s/status/src/index.rs:349]: nit, TIOLI: starting to wonder if we ought to have an `Index::reconcile_if_leader` or something that just does this...

**Review by hawkw — APPROVED** [review]

**Pull request merged**